components/playlist/browser/playlist_service.cc

```diff
@@ -0,0 +1,575 @@
+    "src='video_file.mp4' type='video/mp4' /></video> <video id='a' autoplay "
+    "style='display:none'><source src='audio_file.m4a' type='audio/mp4' "
+    "/></video>";
+
+int DoGenerateHTMLFileOnIOThread(const base::FilePath& html_file_path) {
+  if (base::PathExists(html_file_path))
+    base::DeleteFile(html_file_path, false);
+
+  base::File html_file(html_file_path,
+                       base::File::FLAG_CREATE | base::File::FLAG_WRITE);
+  if (!html_file.IsValid())
+    return -1;
+
+  html_file.WriteAtCurrentPos(kHTMLTemplate, 322 /*kHTMLTemplate.length()*/);
+  return 0;
+}
+
+}  // namespace
+
+PlaylistService::PlaylistService(content::BrowserContext* context)
+    : context_(context),
+      base_dir_(context->GetPath().Append(kBaseDirName)),
+      url_loader_factory_(
+          content::BrowserContext::GetDefaultStoragePartition(context)
+              ->GetURLLoaderFactoryForBrowserProcess()),
+      prefs_(user_prefs::UserPrefs::Get(context)),
+      weak_factory_(this) {
+  content::URLDataSource::Add(
+      context,
+      std::make_unique<PlaylistDataSource>(this));
+
+  // TODO(pilgrim) dynamically set file extensions based on format
+  // (may require changes to youtubedown parser)
+  video_media_file_downloader_.reset(new PlaylistMediaFileDownloader(
+      this, context_, FILE_PATH_LITERAL("video_source_files"),
+      FILE_PATH_LITERAL("video_file.mp4"), kPlaylistVideoMediaFilePathKey,
+      kPlaylistCreateParamsVideoMediaFilesPathKey));
+  audio_media_file_downloader_.reset(new PlaylistMediaFileDownloader(
+      this, context_, FILE_PATH_LITERAL("audio_source_files"),
+      FILE_PATH_LITERAL("audio_file.m4a"), kPlaylistAudioMediaFilePathKey,
+      kPlaylistCreateParamsAudioMediaFilesPathKey));
+
+  CleanUp();
+}
+
+PlaylistService::~PlaylistService() = default;
+
+void PlaylistService::NotifyPlaylistChanged(
+    const PlaylistChangeParams& params) {
+  VLOG(2) << __func__ << ": params="
+          << PlaylistChangeParams::GetPlaylistChangeTypeAsString(
+                 params.change_type);
+
+  for (PlaylistServiceObserver& obs : observers_)
+    obs.OnPlaylistItemStatusChanged(params);
+}
+
+void PlaylistService::AddPlaylistToMediaFileGenerationQueue(
+    const std::string& id) {
+  const base::Value* playlist_info =
+      prefs_->Get(kPlaylistItems)->FindDictKey(id);
+  if (!playlist_info) {
+    LOG(ERROR) << __func__ << ": Invalid playlist id for recover: " << id;
+    return;
+  }
+
+  VLOG(2) << __func__;
+  // Add to pending jobs
+  pending_media_file_creation_jobs_.push(playlist_info->Clone());
+
+  // If either media file controller is generating a playlist media file,
+  // delay the next playlist generation. It will be triggered when the current
+  // one is finished.
+  if (!video_media_file_downloader_->in_progress() &&
+      !audio_media_file_downloader_->in_progress()) {
+    GenerateMediaFiles();
+  }
+}
+
+void PlaylistService::GenerateMediaFiles() {
```

Comment:
can we maybe be extract media files generation to a separate abstraction? Just to keep the service size reasonable. We could encapsulate `video_media_file_downloader_` `audio_media_file_downloader_` and `pending_media_file_creation_jobs_`

Reply:
`PlaylistMediaFileDownloadManager` is introduced. It controls both downloader and job queue.